### ISSUE TITLE
twisted.internet.test.reactormixins.ReactorBuilder.runReactor leaks state on glib2 and gtk2

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3370]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3370|
|Type|defect|
|Created|2008-07-28 14:59:06Z|
|Branch|https://github.com/twisted/twisted/tree/reactorbuilder-cleanup-more-3370|

It seems glib2 and gtk2 reactors live forever and can interfer with later tests.  `ReactorBuilder` needs to do some extra work to clean these up, perhaps by cancelling all their outstanding `DelayedCall`s.

<details><summary>Searchable metadata</summary>

```
trac-id__3370 3370
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__branches_reactorbuilder_cleanup_more_3370 branches/reactorbuilder-cleanup-more-3370
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1217257146000000 1217257146000000
changetime__1217348881000000 1217348881000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [24405]) Branching to 'reactorbuilder-cleanup-more-3370'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

Okay, should be fixed.  You can observe the problem by running `trial -u twisted.internet.test.test_tls` and letting it go for two minutes.  It will die right around the two minute mark.  It should run indefinitely with the changes in the branch.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_3"></a>|@washort set owner to @exarkun|
|-|-|

I get the same results here. Looks good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [24419]) Merge reactorbuilder-cleanup-more-3370

Author: exarkun
Reviewer: washort
Fixes: [#3370](https://github.com/twisted/twisted/issues/3370)

Add code to cancel all pending timed events to the reactor
cleanup code in `twisted.internet.test.reactormixins` so
that they don't remain to interfere with future tests.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
